src/plugin.rs

```diff
@@ -0,0 +1,218 @@
+use std::str::FromStr;
+use std::sync::Arc;
+
+use pyo3::{exceptions::PyAttributeError, prelude::*, types::PyType};
+
+use crate::{DecryptError, EncryptError, IdentityError, RecipientError};
+
+/// Hack, because the orphan rule would prevent us from deriving a
+/// foreign trait on a foreign object. Instead, define a newtype.
+///
+/// Inner type is PyAny, because we do duck-typing at runtime, and
+/// declaring a protocol in the type stubs.
+#[derive(Clone)]
+pub(crate) struct PyCallbacks(Py<PyAny>);
+
+impl PyCallbacks {
+    fn new(inner: Py<PyAny>, py: Python<'_>) -> PyResult<Self> {
+        {
+            let inner = inner.as_ref(py);
+
+            for method_name in [
+                "display_message",
+                "confirm",
+                "request_public_string",
+                "request_passphrase",
+            ] {
+                // I would use `hasattr`, but it doesn't throw, and
+                // constructing pretty AttributeErrors is incredibly
+                // hard.
+                //
+                // XXX: Should we even throw it? Better ask for
+                // forgiveness than permission, and we're panicking
+                // when things get called. And static typecheckers
+                // will catch a protocol violation.
+                inner.getattr(method_name)?;
+            }
```

Comment:
I'm good with this -- it's a bit of a hack, but IMO we might as well enforce presence here rather than cause unnecessary panics below.